silence/silence.go

```diff
@@ -544,104 +540,179 @@ func validateSilence(s *pb.Silence) error {
 	if s.EndsAt.Before(s.StartsAt) {
 		return errors.New("end time must not be before start time")
 	}
-	if s.UpdatedAt.IsZero() {
-		return errors.New("invalid zero update timestamp")
-	}
 	return nil
 }
 
-// cloneSilence returns a shallow copy of a silence.
-func cloneSilence(sil *pb.Silence) *pb.Silence {
-	s := *sil
-	return &s
+// Expire the silence with the given ID immediately.
+func (s *Silences) Expire(id string) error {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	return s.expire(id)
 }
 
-func (s *Silences) getSilence(id string) (*pb.Silence, bool) {
-	msil, ok := s.st[id]
-	if !ok {
-		return nil, false
+// Set inserts a new silence or updates an existing silence. It inserts a
+// silence when the ID is missing or a silence with the ID does not exist.
+// If updating an existing silence, and the update changes how alerts are
+// matched or the time window in which the silence is active (excluding
+// time extensions), the existing silence is expired and replaced with a
+// new silence with a different ID.
+//
+// When updating an existing silence, the existing silence must be cloned
+// using Clone() before making any modifications to the silence. This
+// prevents situations where the stored silence is modified because both
+// pointers reference the same memory. If Clone() is not used before
```

Comment:
`Clone()` is required otherwise functions like `canUpdate` break. Again, no changes here, this is just documentation.